### PR TITLE
Release/next wixhelper

### PR DESCRIPTION
**Pull request by cthangn**
Add Wix to FAKE 5

### CODE REVIEW

File: src/app/Fake.Windows.Wix/Wix.fs [hunk 1]
```diff
@@ -0,0 +1,2008 @@
+﻿/// Contains tasks to create msi installers using the [WiX toolset](http://wixtoolset.org/)
+
+module Fake.Windows.Wix
```
Comment (matthid) [inline src/app/Fake.Windows.Wix/Wix.fs]: please add `[<RequireQualifiedAccess>] ` if reasonable

File: src/app/Fake.Windows.Wix/Wix.fs [hunk 2]
```diff
@@ -0,0 +1,2008 @@
...
+open System.Text
+open Fake.IO
+open Fake.IO.FileSystemOperators
+open Fake.IO.Globbing.Operators
+
+
+let mutable internal fileCount = 0
+let mutable internal dirs = Dictionary()
+
+let getDirName dir =
```
Comment (matthid) [inline src/app/Fake.Windows.Wix/Wix.fs]: internal?

File: src/app/Fake.Windows.Wix/Wix.fs [hunk 3]
```diff
@@ -0,0 +1,2008 @@
...
+    | true, n -> 
+        dirs.[dir] <- n + 1
+        dir + n.ToString()
+    | _ -> 
+        dirs.[dir] <- 1
+        dir
+
+let mutable internal compRefs = Dictionary()
+
+let getCompRefName compRef =
```
Comment (matthid) [inline src/app/Fake.Windows.Wix/Wix.fs]: internal?

File: src/app/Fake.Windows.Wix/Wix.fs [hunk 4]
```diff
@@ -0,0 +1,2008 @@
...
+    | true, n -> 
+        compRefs.[compRef] <- n + 1
+        compRef + n.ToString()
+    | _ -> 
+        compRefs.[compRef] <- 1
+        compRef
+
+let mutable internal comps = Dictionary()
+
+let getCompName comp =
```
Comment (matthid) [inline src/app/Fake.Windows.Wix/Wix.fs]: internal?

File: src/app/Fake.Windows.Wix/Wix.fs [hunk 5]
```diff
@@ -0,0 +1,2008 @@
...
+            (Seq.fold(fun acc (key, value) -> acc + sprintf " %s=\"%s\"" key value) "" (w.createAttributeList())) 
+
+let ServiceDependencyDefaults =
+    {
+        Id = ""
+        Group = None
+    }
+
+/// Use this for generating service dependencies
+let generateServiceDependency (setParams : ServiceDependency -> ServiceDependency) =
```
Comment (matthid) [inline src/app/Fake.Windows.Wix/Wix.fs]: internal?

File: src/app/Fake.Windows.Wix/Wix.fs [hunk 6]
```diff
@@ -0,0 +1,2008 @@
...
+    member w.createAttributeList () =
+        seq {           
+            yield ("Id", w.Id)           
+            if w.Group.IsSome then yield ("Group", w.Group.Value.ToString())
+        }
+    override w.ToString() =
+        sprintf "<ServiceDependency%s />"
+            (Seq.fold(fun acc (key, value) -> acc + sprintf " %s=\"%s\"" key value) "" (w.createAttributeList())) 
+
+let ServiceDependencyDefaults =
```
Comment (matthid) [inline src/app/Fake.Windows.Wix/Wix.fs]: internal?

File: src/app/Fake.Windows.Wix/Wix.fs [hunk 7]
```diff
@@ -0,0 +1,2008 @@
...
+        RebootMessage = ""
+        ResetPeriodInDays = 0
+        RestartServiceDelayInSeconds = 0
+        SecondFailureActionType = NoneAction
+        ServiceName = ""
+        ThirdFailureActionType = NoneAction
+    }
+
+/// Use this for generating service configs
+let generateServiceConfig (setParams : ServiceConfig -> ServiceConfig) =
```
Comment (matthid) [inline src/app/Fake.Windows.Wix/Wix.fs]: internal?

File: src/app/Fake.Windows.Wix/Wix.fs [hunk 8]
```diff
@@ -0,0 +1,2008 @@
...
+                yield ("SecondFailureActionType", w.SecondFailureActionType.ToString())                
+                if not (String.IsNullOrWhiteSpace w.ServiceName) then
+                    yield ("ServiceName", w.ServiceName)                
+                yield ("ThirdFailureActionType", w.ThirdFailureActionType.ToString())
+            }
+    override w.ToString() = 
+        sprintf "<ServiceConfig xmlns=\"http://schemas.microsoft.com/wix/UtilExtension\" %s/>" 
+            (Seq.fold(fun acc (key, value) -> acc + sprintf " %s=\"%s\"" key value) "" (w.createAttributeList())) 
+
+let ServiceConfigDefaults =
```
Comment (matthid) [inline src/app/Fake.Windows.Wix/Wix.fs]: internal?

File: src/app/Fake.Windows.Wix/Wix.fs [hunk 9]
```diff
@@ -0,0 +1,2008 @@
...
+                | Never -> ()
+                | _ -> yield ("Stop", w.Stop.ToString())
+                yield ("Wait", w.Wait.ToString())
+            }
+    override w.ToString() = 
+        sprintf "<ServiceControl%s/>" 
+            (Seq.fold(fun acc (key, value) -> acc + sprintf " %s=\"%s\"" key value) "" (w.createAttributeList()))             
+
+/// Defaults for service control element
+let ServiceControlDefaults =
```
Comment (matthid) [inline src/app/Fake.Windows.Wix/Wix.fs]: internal?

File: src/app/Fake.Windows.Wix/Wix.fs [hunk 10]
```diff
@@ -0,0 +1,2008 @@
...
+        Id = "ServiceControl"
+        Name = "Service"
+        Remove = Both
+        Start = Install
+        Stop = Both
+        Wait = Yes
+    }
+
+/// Use this for generating service controls
+let generateServiceControl (setParams : ServiceControl -> ServiceControl) =
```
Comment (matthid) [inline src/app/Fake.Windows.Wix/Wix.fs]: internal?

File: src/app/Fake.Windows.Wix/Wix.fs [hunk 11]
```diff
@@ -0,0 +1,2008 @@
...
+        AllowDowngrades : YesOrNo
+        /// The message displayed if users try to install a product with a lower version number when a product with a higher version is installed. Used only when AllowDowngrades is no (the default). 
+        DowngradeErrorMessage : string
+    }
+    override w.ToString() =
+        let downgradeErrorMessage =
+            match w.AllowDowngrades with
+            | Yes -> ""
+            | No ->  " DowngradeErrorMessage=\"" + w.DowngradeErrorMessage + "\""
+        "<MajorUpgrade Schedule=\"" + w.Schedule.ToString() +
```
Comment (matthid) [inline src/app/Fake.Windows.Wix/Wix.fs]: remove this function?

**Comment by kblohm** [comment]
Probably not too important, but shouldn't this go under Fake.Installer? There is also an old Squirrel-Helper and it might make it harder to find all of these.

File: src/app/Fake.Windows.Wix/Wix.fs (hunk 11 above)
Comment (cthangn) [reply src/app/Fake.Windows.Wix/Wix.fs]: I don't quite understand, which function I should remove here. Can you please clarify?

**Comment by matthid** [comment]
I tend to agree with @kblohm especially since we have the `Fake.Installer` namespace already. So can you move this to `Fake.Installer.Wix` please?

The second thing is that there are still a lot "helper" functions which are public and I don't think they should be. However, I never used this module myself so it might be that they are indeed useful for using this module. In particular the `*.Defaults` field are usually `internal`/`private` in fake 5. But also others like `getWixFileTag` and `getFilesAsWiXString` look a lot like internal helpers and not used from users via `Wix.getFilesAsWiXString`. 

It is easier for us to make things public later on than the other way around. @cthangn can you please go over the functions again with that mindset (sorry I didn't mark every place in the first review because there were too many).

Sorry for the late reply.

**Comment by cthangn** [comment]
Thank you for the clarification. I have moved `Wix` under `Fake.Installer`. I have marked all `*Defaults` internal, which makes sense. 

About `getFilesAsWiXString` there is an example in the code how to use it. 

About those `generate*` functions, we are using some of them even in our script and they are helpers for users to create those required structures with overriding the `*Defaults`, so I think they should be `public`. Maybe there are some, which can be `internal/private`, but I don't have such extensive knowledge of this module to decide which ones.

**Comment by matthid** [comment]
>  but I don't have such extensive knowledge of this module to decide which ones.

Me neither, but wheen in doubt we should make internal/private and wait for people to complain then we can defer designing a proper API for such functions to a later point... Because for all public functions we need to make sure they align with the new API guideline.

**Comment by matthid** [comment]
Ok sorry for the back and forth. Good job looks really neat now!

**Pull request merged**